lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -58,6 +59,34 @@ static ParseResult parseCircuitOp(OpAsmParser &parser, OperationState &result) {
   return success();
 }
 
+static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
+  StringRef main = circuit.name();
+
+  // Check that the circuit must have a "main" module (top module)
+  if (main.empty()) {
+    circuit.emitOpError();
+    return failure();
+  }
+
+  StringSet<> *moduleNames = new StringSet<>();
+
+  for (auto &op : *circuit.getBody()) {
+    if (auto module = dyn_cast<FModuleOp>(op)) {
+      if (op.getAttrOfType<StringAttr>(SymbolTable::getSymbolAttrName())) {
+        moduleNames->insert(module.getName());
+      }
+    }
+  }
+
+  // Check that a module matching the "main" module exists in the circuit
+  if (moduleNames->find(main) == moduleNames->end()) {
+    circuit.emitOpError("must contain one module that matches main name ('" + main + "')");
```

Comment:
Plz fit to 80 columns

Reply:
Also, nit: I'd recommend dropping the parens around the name, the single quotes are enough.

Reply:
80-column applied in a later commit via `clang-format`. Error message updated.